File: LC/python/Validate Binary Search Tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a boolean
    def isValidBST(self, root):
        def dfs(root,minV,maxV):
            if root == None:
                return True
            if root.val<=minV or root.val>=maxV:
                return False
            return dfs(root.left,minV,root.val) and dfs(root.right,root.val,maxV)
            
        return dfs(root,float("-inf"),float("inf"))

File: LC/java/Validate Binary Search Tree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    private boolean first = true;
    private int last = Integer.MIN_VALUE;
    public boolean isValidBST(TreeNode root) {
       
       if(root == null){
           return true;
       } 
       
       if(!isValidBST(root.left)){
           return false;
       }
       
       if(!first && last>=root.val)
       {
           return false;
       }
       
       first = false;
       last = root.val;
       
       if(!isValidBST(root.right)){
           return false;
       }
       return true;
    }
}